### ISSUE TITLE
[Feature Request] Implement Recurrent SAC

### ISSUE THREAD

**Issue by masterdezign** [Stable-Baselines-Team/stable-baselines3-contrib#201]
### 🚀 Feature

Hi! 

I would like to implement a recurrent soft actor-critic. Is it a sensible contribution?

### Motivation

I actually need this algorithm in my projects.

### Pitch

The sb3 ecosystem would benefit from yet another algorithm. As a new contributor, I might need a little guidance though.

### Alternatives

An alternative would be another off-policy algorithm using LSTM.

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/issues) in the repo
- [X] If I'm requesting a new feature, I have proposed alternatives

**Comment by araffin** [comment]
Hello,
this would be definitely a good addition to SB3 contrib.

Make sure to read the contributing guide carefully.
You might have a look at R2D2 paper  (https://paperswithcode.com/method/r2d2) and https://github.com/zhihanyang2022/off-policy-continuous-control.

For benchmarking, best would be to use the "NoVel" env that are available in the RL Zoo (see https://wandb.ai/sb3/no-vel-envs/reports/PPO-vs-RecurrentPPO-aka-PPO-LSTM-on-environments-with-masked-velocity-SB3-Contrib---VmlldzoxOTI4NjE4).

**Comment by masterdezign** [comment]
Thanks for the references. I will check them out and come back.

**Comment by masterdezign** [comment]
Just a quick update: I plan to do this by the end of 2023 when I have some free time. Currently I have three higher priority projects.

**Comment by masterdezign** [comment]
Status update:

1. I've checked the resources that you provided, thanks a lot. I find the [code](https://github.com/zhihanyang2022/off-policy-continuous-control) to be nicely written and quite easy to understand.
2. I managed to solve PendulumNoVel-v1 from `rl_zoo3==2.1.0` with [RSAC](https://github.com/zhihanyang2022/off-policy-continuous-control).
3. However, I have trouble solving `MountainCarContinuousNoVel-v0` and `LunarLanderContinuousNoVel-v2` using the code above with different [configurations](https://github.com/zhihanyang2022/off-policy-continuous-control/tree/pub/offpcc/configs).
4. Therefore, I may need to resort to modify the algorithm (such as e.g. using the same LSTM state by the actor and critics, using overlapping segments, etc.).

**Comment by masterdezign** [comment]
![rsac_s-241](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/assets/756621/e218568e-c0b7-41fe-94e3-1491dd44ebe5)


So I've got this result on `LunarLanderContinuousNoVel-v2` using shared LSTM state ([rsac_s](https://github.com/zhihanyang2022/off-policy-continuous-control)) and the following configuration

```
capacity = 1000  # 1M / 1000 = 1000
batch_size = 10
segment_len = 50

num_epochs = 500
num_steps_per_epoch = 10000
update_after = 10000
num_test_episodes_per_epoch = 10
```

Shall we consider this a successful run? If so, perhaps now this `rsac_s` architecture can be implemented in sb3-contrib.

**Comment by araffin** [comment]
Hello,
thanks for reporting the updated results =).
Do you have a diagram to share for RSAC vs RSAC_s maybe? (that would make things easier to discuss)

Di you also manage to solve the mountain car problem?

**Comment by masterdezign** [comment]
> Di you also manage to solve the mountain car problem?

I believe, yes. Let me render the env to verify since rewards are not the same for MountainCarContinuousNoVel-v0 (continuous action space) and MountainCar-v0 (discrete action space).

**Comment by masterdezign** [comment]
Loosely speaking, here they are:

```

           RSAC                        RSAC_S

     ┌─────┐    ┌─────┐               ┌─────┐
     │ RNN │    │ RNN │             ┌─┤ RNN │..
     └──┬──┘    └──┬──┘             │ └─────┘ .
        │          │                │         .
        │          │                │         .
    ┌───┴───┐  ┌───┴────┐       ┌───┴───┐  ┌────────┐
    │ Actor │  │ Critic │       │ Actor │  │ Critic │
    └───────┘  └────────┘       └───────┘  └────────┘

```

As you can see, RSAC_S share the RNN state between the actor and the critic, but only actor can change the RNN state. Whereas in RSAC actor and critics have their own RNN states.

**Comment by araffin** [comment]
> As you can see, RSAC_S share the RNN state between the actor and the critic, but only actor can change the RNN state. Whereas in RSAC actor and critics have their own RNN states.

thanks, similar to what is implemented for PPO:
https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/588c6bdaeaa118a075162eddcd77c753d880bee2/sb3_contrib/common/recurrent/policies.py#L238-L247

**Comment by masterdezign** [comment]
Update, I just rendered `MountainCarContinuousNoVel-v0` and it is **not** solved yet. I don't quite understand why the total reward is different between the original `MountainCar-v0` env and this one. Therefore, I need to check `MountainCarContinuousNoVel-v0` in details.

**Comment by araffin** [comment]
i can help you with that, the continuous version has a deceptive reward and need quite some exploration

**Comment by masterdezign** [comment]
Indeed, having use_sde=True seems helping to solve `MountainCarContinuous-v0` environment. I am curious which gSDE ingredient does exactly help. Also, I may want to try nearby hyperparameters to see how robust is gSDE contribution or we are just lucky.

**Comment by araffin** [comment]
> I am curious which gSDE ingredient does exactly help.

The consistent exploration. To solve this task, you need to build-up momentum, having a bang-bang like strategy is one way (it is discuss a bit longer in the first version of the paper: https://arxiv.org/pdf/2005.05719v1.pdf).

> Edit: I also tried nearby hyperparameters and indeed gSDE contribution seems to be non-negligible.

I did a full hyperparameters search and with gSDE many are working (more than half of the tested configurations): https://github.com/DLR-RM/rl-baselines3-zoo/blob/sde/logs/report_sde_MountainCarContinuous-v0_500-trials-50000-tpe-median_1581693633.csv

**Comment by masterdezign** [comment]
I am currently checking the two strategies for RNN state initialization, proposed in R2D2 paper (store state and burn-in).

**Comment by masterdezign** [comment]
So far I've got this: recurrent replay buffer with overlapping chunks supporting SB3 interface. I also wrote a specification (test) to reduce surprise in the future.

https://gist.github.com/masterdezign/47b3c6172dd1624bb9a7ef23cbc79c8c

**Comment by masterdezign** [comment]
Hi! I didn't obtain good results and then I had to put the project on hold. I plan to restart working on it starting from tomorrow.

**Comment by masterdezign** [comment]
Found these related discussions in Ray project:
* https://github.com/ray-project/ray/pull/49680
* https://github.com/ray-project/ray/pull/48116
